File: tests/xcalablemp/global-view/gmove/F/gmove_G2L_16.f90
program gmove_G2L_16

  character*25 tname
  call gmove_G2L_1a4t_bc(tname)
  call gmove_G2L_4a4t_bc(tname)

end program

subroutine gmove_G2L_1a4t_bc(tname)

character(*) tname
integer,parameter :: n=8
integer a(n,n,n,n), b(n,n,n,n)
integer xmp_node_num
!$xmp nodes p(2,2,2,2)
!$xmp template tx(n,n,n,n)
!$xmp distribute tx(cyclic(2),cyclic(2),cyclic(2),cyclic(2)) onto p
!$xmp align a(*,*,*,i3) with tx(*,*,*,i3)

!$xmp loop (i3) on tx(*,*,*,i3)
do i3=1,n
  do i2=1,n
    do i1=1,n
      do i0=1,n
        a(i0,i1,i2,i3)=i0+i1+i2+i3
      end do
    end do
  end do
end do

do i3=1,n
  do i2=1,n
    do i1=1,n
      do i0=1,n
        b(i0,i1,i2,i3)=0
      end do
    end do
  end do
end do

!$xmp gmove
b(2:n,2:n,2:n,2:n)=a(2:n,2:n,2:n,2:n)

ierr=0
do i3=2,n
  do i2=2,n
    do i1=2,n
      do i0=2,n
        ierr=ierr+abs(b(i0,i1,i2,i3)-i0-i1-i2-i3)
      end do
    end do
  end do
end do

!$xmp reduction (max:ierr)
tname="gmove_G2L_1a4t_bc"
call chk_int(tname, ierr)

end subroutine

subroutine gmove_G2L_4a4t_bc(tname)

character(*) tname
integer i
integer,parameter :: n=8
integer a(n,n,n,n),b(n,n,n,n)
integer xmp_node_num
!$xmp nodes p(2,2,2,2)
!$xmp template tx(n,n,n,n)
!$xmp distribute tx(cyclic(2),cyclic(2),cyclic(2),cyclic(2)) onto p
!$xmp align a(i0,i1,i2,i3) with tx(i0,i1,i2,i3)

!$xmp loop (i0,i1,i2,i3) on tx(i0,i1,i2,i3)
do i3=1,n
  do i2=1,n
    do i1=1,n
      do i0=1,n
        a(i0,i1,i2,i3)=i0+i1+i2+i3
      end do
    end do
  end do
end do

do i3=1,n
  do i2=1,n
    do i1=1,n
      do i0=1,n
        b(i0,i1,i2,i3)=0
      end do
    end do
  end do
end do

!$xmp gmove
b(2:5,2:5,2:5,2:5)=a(5:8,5:8,5:8,5:8)

ierr=0
do i3=2,5
  do i2=2,5
    do i1=2,5
      do i0=2,5
        ierr=ierr+abs(b(i0,i1,i2,i3)-(i0+3)-(i1+3)-(i2+3)-(i3+3))
      end do
    end do
  end do
end do

!$xmp reduction (max:ierr)
tname="gmove_G2L_4a4t_bc"
call chk_int(tname, ierr)

end subroutine

File: tests/XMP/global-view/gmove/C-Round/gmove_G2L_16.c
#define NAMELEN 25
#define N 8
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

extern int chk_int(char name[], int ierr);

void gmove_G2L_1a4t_bc(){

char name[NAMELEN]="gmove_G2L_1a4t_bc";
int n=N;
double a[n][n][n][n];
#pragma xmp nodes p(2,2,2,2)
#pragma xmp template tx(0:n-1,0:n-1,0:n-1,0:n-1)
#pragma xmp distribute tx(cyclic(2),cyclic(2),cyclic(2),cyclic(2)) onto p
#pragma xmp align a[*][*][*][i] with tx(*,*,*,i)

  int i0,i1,i2,i3,ierr;
  double b[N][N][N][N],err;

  for(i3=0;i3<n;i3++){
    for(i2=0;i2<n;i2++){
      for(i1=0;i1<n;i1++){
#pragma xmp loop (i0) on tx(*,*,*,i0)
        for(i0=0;i0<n;i0++){
          a[i3][i2][i1][i0]=i3+i2+i0+i1+1;
        }
      }
    }
  }

  for(i3=0;i3<n;i3++){
    for(i2=0;i2<n;i2++){
      for(i1=0;i1<n;i1++){
        for(i0=0;i0<n;i0++){
          b[i3][i2][i1][i0]=0.0;
        }
      }
    }
  }

#pragma xmp gmove
  b[1:n-1][1:n-1][1:n-1][1:n-1]=a[1:n-1][1:n-1][1:n-1][1:n-1];

  err=0.0;
  for(i3=1;i3<n;i3++){
    for(i2=1;i2<n;i2++){
      for(i1=1;i1<n;i1++){
        for(i0=1;i0<n;i0++){
          err=err+fabs(b[i3][i2][i1][i0]-i3-i2-i0-i1-1);
        }
      }
    }
  }

#pragma xmp reduction (MAX:err)
  ierr=err;
  chk_int(name, ierr);

}

void gmove_G2L_4a4t_bc(){

char name[NAMELEN]="gmove_G2L_4a4t_bc";
int n=N;
int a[n][n][n][n];
#pragma xmp nodes p(2,2,2,2)
#pragma xmp template tx(0:n-1,0:n-1,0:n-1,0:n-1)
#pragma xmp distribute tx(cyclic(2),cyclic(2),cyclic(2),cyclic(2)) onto p
#pragma xmp align a[i0][i1][i2][i3] with tx(i0,i1,i2,i3)

  int i0,i1,i2,i3,b[N][N][N][N],ierr=0;

#pragma xmp loop (i0,i1,i2,i3) on tx(i0,i1,i2,i3)
  for (i3=0;i3<n;i3++){
    for (i2=0;i2<n;i2++){
      for (i1=0;i1<n;i1++){
        for (i0=0;i0<n;i0++){
          a[i0][i1][i2][i3]=(i0+1)+(i1+1)+(i2+1)+(i3+1);
        }
      }
    }
  }

  for (i3=0;i3<n;i3++){
    for (i2=0;i2<n;i2++){
      for (i1=0;i1<n;i1++){
        for (i0=0;i0<n;i0++){
          b[i0][i1][i2][i3]=0;
        }
      }
    }
  }

#pragma xmp gmove
  b[1:4][1:4][1:4][1:4]=a[4:4][4:4][4:4][4:4];

  for (i3=1;i3<5;i3++){
    for (i2=1;i2<5;i2++){
      for (i1=1;i1<5;i1++){
        for (i0=1;i0<5;i0++){
          ierr=ierr+abs(b[i0][i1][i2][i3]-(i0+4)-(i1+4)-(i2+4)-(i3+4));
        }
      }
    }
  }

#pragma xmp reduction (+:ierr)
  chk_int(name, ierr);

}

int main(){

  gmove_G2L_1a4t_bc();
  gmove_G2L_4a4t_bc();

  return 0;

}
